mmcv/ops/csrc/pytorch/cpu/correlation.cpp

```diff
@@ -0,0 +1,188 @@
+                                         scalar_t *dst, int kH, int kW,
+                                         int dilationH, int dilationW, int u,
+                                         int v, int shiftU, int shiftV) {
+  const int C = input1.size(0);
+  const int iH = input1.size(1);
+  const int iW = input1.size(2);
+  for (int c = 0; c < C; ++c) {
+    for (int i = 0; i < kH; ++i) {
+      int i1 = u + i * dilationH;
+      int i2 = i1 + shiftU;
+      if WITHIN_BOUNDS (i1, i2, iH, iH) {
+        for (int j = 0; j < kW; ++j) {
+          int j1 = v + j * dilationW;
+          int j2 = j1 + shiftV;
+          if WITHIN_BOUNDS (j1, j2, iW, iW) {
+            scalar_t v1 = input1[c][i1][j1];
+            scalar_t v2 = input2[c][i2][j2];
+            *dst += v1 * v2;
+          }
+        }
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+static void correlation_backward_cpu_kernel(
+    TensorAccessor<scalar_t, 3> input1, TensorAccessor<scalar_t, 3> gradInput1,
+    TensorAccessor<scalar_t, 3> input2, TensorAccessor<scalar_t, 3> gradInput2,
+    scalar_t &gradOutput, int kH, int kW, int dilationH, int dilationW, int u,
+    int v, int shiftU, int shiftV) {
+  const int C = input1.size(0);
+  const int iH = input1.size(1);
+  const int iW = input1.size(2);
+
+  for (int c = 0; c < C; ++c) {
+    for (int i = 0; i < kH; ++i) {
+      int i1 = u + i * dilationH;
+      int i2 = i1 + shiftU;
+      if WITHIN_BOUNDS (i1, i2, iH, iH) {
+        for (int j = 0; j < kW; ++j) {
+          int j1 = v + j * dilationW;
+          int j2 = j1 + shiftV;
+          if WITHIN_BOUNDS (j1, j2, iW, iW) {
+            scalar_t v1 = input1[c][i1][j1];
+            scalar_t v2 = input2[c][i2][j2];
+            gradInput2[c][i2][j2] += gradOutput * v1;
+            gradInput1[c][i1][j1] += gradOutput * v2;
+          }
+        }
+      }
+    }
+  }
+}
+
+void correlationForwardCPUKernelLaucher(const Tensor input1,
+                                        const Tensor input2, Tensor output,
+                                        int kH, int kW, int patchH, int patchW,
+                                        int padH, int padW, int dilationH,
+                                        int dilationW, int dilation_patchH,
+                                        int dilation_patchW, int dH, int dW) {
+  const auto batch_size = input1.size(0);
+  const auto iH = input1.size(2);
+  const auto iW = input1.size(3);
+  const int patchRadH = (patchH - 1) / 2;
+  const int patchRadW = (patchW - 1) / 2;
+  const int dilatedKH = (kH - 1) * dilationH + 1;
+  const int dilatedKW = (kW - 1) * dilationW + 1;
+
+  const auto oH = (iH + 2 * padH - dilatedKH) / dH + 1;
+  const auto oW = (iW + 2 * padW - dilatedKW) / dW + 1;
+  // auto output = at::zeros({batch_size, patchH, patchW, oH, oW},
+  // input1.options());
+
+  int n, ph, pw, h, w;
+#pragma omp parallel for private(n, ph, pw, h, w) collapse(2)
```

Comment:
Should we add compiler flags to enable openmp?
I got warning when compile the ops 
```dotnetcli
correlation.cpp:90: warning: ignoring #pragma omp parallel [-Wunknown-pragmas]
```

Reply:
emmmm, yes. and there should be some dependencies,  maybe remove it is a better choice, if performance is not important.